### PR TITLE
[Sema] SR-2327: Improve error message when using 'break' inside 'guar…

### PR DESCRIPTION
**Pull request by MnO2**
#### What's in this pull request?

Improve the error message what SR-2327 has described.

With

```
do {
    guard 1 == 2 else { break }
}
```

prompt the more helpful error message of `error: unlabeled 'break' is only allowed inside a loop or switch, a labeled break is required to exit an if or do`
#### Resolved bug number: ([SR-2327](https://bugs.swift.org/browse/SR-2327))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…d' inside 'do' or 'if'

### CODE REVIEW

**Comment by CodaFi** [comment]
Very cool!

@swift-ci please smoke test OS X platform.

File: lib/Sema/TypeCheckStmt.cpp [hunk 1]
```diff
@@ -767,7 +767,8 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
       // statement, produce a more specific error.
       if (S->getTargetName().empty() && !ActiveLabeledStmts.empty() &&
           (isa<IfStmt>(ActiveLabeledStmts.back()) ||
-           isa<DoStmt>(ActiveLabeledStmts.back())))
+           isa<DoStmt>(ActiveLabeledStmts.back()) ||
+           isa<GuardStmt>(ActiveLabeledStmts.back())))
```
Comment (rintaro) [inline lib/Sema/TypeCheckStmt.cpp]: Does this work for this?
```swift
func fn(a: Int) {
  guard a < 1 else {
    break
  }
}
```
In this case, IMO, the message should be normal `'break' is only allowed inside a loop, if, do, or switch`.

Comment (tkremenek) [inline lib/Sema/TypeCheckStmt.cpp]: Can we put `if`, `do`, in single quotes, just like `break`?

Comment (MnO2) [inline lib/Sema/TypeCheckStmt.cpp]: @rintaro I've updated the code to handle the case and added your example into the test cases.

@tkremenek I prefer to submit another pull request to change the format of diagnosis message. I am also in favor of quoting `if` and `do`, but touching the message in this commit would make some noises on the logic change behind this commit.

Comment (tkremenek) [inline lib/Sema/TypeCheckStmt.cpp]: @rintaro That works for me.

**Comment by xwu** [comment]
One suggestion: either in this commit or in the follow-up, probably best to change the `guard_body_must_not_fallthrough` diagnostic message from...

'guard' body may not fall through, consider using 'return' or 'break'

...to...

'guard' body may not fall through, consider using 'return' or **labeled** 'break'

File: lib/Sema/TypeCheckStmt.cpp [hunk 2]
```diff
@@ -765,10 +765,23 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+        if (isa<GuardStmt>(ActiveLabeledStmts.back())) {
+          if (ActiveLabeledStmts.size() >= 2) {
+            auto parent = ActiveLabeledStmts.end()[-2];
+            if (isa<IfStmt>(parent) ||
+                isa<DoStmt>(parent)) {
+              diagid = diag::unlabeled_break_outside_loop;
+            }
+          }
+        }
+      }
```
Comment (rintaro) [inline lib/Sema/TypeCheckStmt.cpp]: Sorry for nit-picking. This doesn't work for
```swift
func fn(x: Int) {
  if x > 0 {
    gurad x < 1 {
      guard x < 2 {
        break
      }
      return
    }
  }
}
```
The rule here is: If `ActiveLabeledStmts` contains `if` or `do`, we can use `unlabeled_break_outside_loop`.
I think, something like
```cpp
if (S->getTargetName().empty() && 
    std::any_of(ActiveLabeledStmts.rbegin(), ActiveLabeledStmts.rend(),
      [&](Stmt *S) -> bool { return isa<IfStmt>(S) || isa<DoStmt>(S); }))
```
should work.

Comment (MnO2) [inline lib/Sema/TypeCheckStmt.cpp]: You are absolutely right, I would change accordingly.

Comment (MnO2) [inline lib/Sema/TypeCheckStmt.cpp]: I've made the change and updated the test cases. :-)

File: lib/Sema/TypeCheckStmt.cpp [hunk 3]
```diff
@@ -765,10 +765,22 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+       
+        if (isa<GuardStmt>(ActiveLabeledStmts.back())) {
+          if (ActiveLabeledStmts.size() >= 2 &&
+              std::any_of(ActiveLabeledStmts.rbegin(), ActiveLabeledStmts.rend(),
+                          [&](Stmt *S) -> bool { return isa<IfStmt>(S) || isa<DoStmt>(S); })) {
+
+            diagid = diag::unlabeled_break_outside_loop;
+          }
+        }
+      }
```
Comment (rintaro) [inline lib/Sema/TypeCheckStmt.cpp]: I meant, this whole `if` block can be replaced with:
```
if (S->getTargetName().empty() && 
    std::any_of(ActiveLabeledStmts.rbegin(), ActiveLabeledStmts.rend(),
      [&](Stmt *S) -> bool { return isa<IfStmt>(S) || isa<DoStmt>(S); }))
  diagid = diag::unlabeled_break_outside_loop;
```
I think we don't need to care about `guard` statement here.
Am I missing something?

**Comment by jrose-apple** [comment]
The intent of "return or break" is to break out of a *loop* being guarded, not to exit the `guard` itself.

**Comment by rintaro** [comment]
@jrose-apple  I don't understand the point. What do you mean?
We can break out of labeled `if` or `do` as well.
```
LABEL: do {
  guard [condition] else {
    break LABEL
  }
  // ...
}
```

**Comment by jrose-apple** [comment]
Sorry, yes, you can. But to have the error message suggest a labeled break (per @xwu) might imply that you can't use it with unlabeled break, which of course you can if there's a loop:

```swift
for person in friends {
  guard person.name != "Jordan" else { break }
}
```

**Comment by xwu** [comment]
(D'oh!)

**Comment by rintaro** [comment]
Ah, make sense. You are replying to @xwu's comment 😄

File: lib/Sema/TypeCheckStmt.cpp (hunk 3 above)
Comment (MnO2) [inline lib/Sema/TypeCheckStmt.cpp]: Oh! duh, (to myself) what were you thinking. Thanks for pointing out. I've stolen your snippet and copied-pasted and pushed. :)

File: lib/Sema/TypeCheckStmt.cpp [hunk 4]
```diff
@@ -765,10 +765,11 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
 
       // If someone is using an unlabeled break inside of an 'if' or 'do'
       // statement, produce a more specific error.
-      if (S->getTargetName().empty() && !ActiveLabeledStmts.empty() &&
-          (isa<IfStmt>(ActiveLabeledStmts.back()) ||
-           isa<DoStmt>(ActiveLabeledStmts.back())))
+      if (S->getTargetName().empty() &&
+          std::any_of(ActiveLabeledStmts.rbegin(), ActiveLabeledStmts.rend(),
+                      [&](Stmt *S) -> bool { return isa<IfStmt>(S) || isa<DoStmt>(S); })) {
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckStmt.cpp]: Nitpick: swiftc style limits lines to 80 columns. Can you reflow this, or use a helper variable?

Comment (MnO2) [inline lib/Sema/TypeCheckStmt.cpp]: Sure, I've inserted a few new lines in between.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
[Linux](https://ci.swift.org/view/Pull%20Request/job/swift-PR-Linux/3460/), [OS X](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/3227/)

**Comment by rintaro** [comment]
Test passed.
The change looks good to me.
Thank you! @MnO2

**Comment by jrose-apple** [comment]
Force-merging. Thanks, Paul!

**Comment by tkremenek** [comment]
@MnO2 please create a pull request for `swift-3.0-branch` to pull this change in there.

**Comment by MnO2** [comment]
I cherry-picked the change and created pull-request against `swift-3.0-branch`: https://github.com/apple/swift/pull/4345
